### ISSUE TITLE
JSVerify is looking for a new maintainer.

### ISSUE THREAD

**Issue by phadej** [jsverify/jsverify#299]
It's quite clear that I don't have enough time to even do the very basic maintenance tasks. Nor I write JavaScript/TypeScript anymore. So, I'm looking for a new maintainer(s).

Unfortunately I don't have anyone in mind to ask directly. IIRC @StoneCypher have been active answering in many issues. Maybe @jvanbruegge know someone interested?

- The job is to make and keep `jsverify` *the* property testing library in JavaScript (with TypeScript typings)
- There aren't any formal requirements, but I don't want to give this library to just anyone either. Based on [npmjs package stats](https://www.npmjs.com/package/jsverify), someone is using it!
- I can send you few [HaskHEL stickers](https://github.com/haskhel/logo/blob/master/haskhel.svg) as a gratitude

You may publicly ask questions in this issue, or you could send them (and "applications") to oleg.grenrus@iki.fi

---

**Q:** What about dependencies?
**A:** `jsverify` indeed depends on a few libraries I also authored. I will transfer them to this organisation. Their maintenance is a part of a deal.

**Comment by dubzzz** [comment]
Hi @phadej,

I personally wrote another framework for property based testing in JavaScript called [fast-check](https://github.com/dubzzz/fast-check).

As the API is close, maybe the alternative could be to suggest people ways to move from one framework to the other.

Contrary to jsverify, DSL is not part of the framework but it comes with other useful features such as:
- ability to shrink `sum`
- no need for a reverse operation in `smap` to have shrinking capabilities
- easier debugging: verbose mode, replay of the minimal counterexample
- command based capabilities - *this point and `sum` not shrinking were the initial reason why I wrote fast-check (initially as a POC)*

Regards ;)

**Comment by StoneCypher** [comment]
I have been active, and I would be happy to pitch in some, but it's worth noting that a few people like @lastmjs and @glebec have made more contributions than I have

**Comment by StoneCypher** [comment]
i am actually kind of taking this seriously though

my opinion is that jsverify is the conceptually strongest built of the property testers, but also, that its documentation state is some of the most difficult

it also has most of its examples written in a different programming language that has little overlap with the target userbase

i feel that jsverify would benefit greatly from two things: a significantly changed documentation culture, and the deployment of a large set of generators, intended to take on something of a similar worldview to `faker`, but with reduction and composition

this would mean giving a lot of examples of The Right Way To Do It &trade; , in documentation

would this seem of interest to you, @phadej ?

**Comment by lastmjs** [comment]
I've moved on to https://github.com/dubzzz/fast-check, perhaps the projects could be merged or effort given to one or the other instead of maintaining two, unless there are some fundamental tradeoffs that can't be reconciled

**Comment by glebec** [comment]
@StoneCypher I've made 0 code contributions to JSVerify though I recently was pretty active in browsing through issues, PRs, and the source code – part of writing [this article series](https://dev.to/glebec/property-testing-with-jsverify-part-i-3one). I definitely have some opinions here and there as to small specific improvements or fixes, but that's a very different level of commitment from being a maintainer (though I'm flattered my name was brought up).

I also have taken a short glance at @dubzzz's library `fast-check` and like it a lot so far. Haven't done a deep dive so I wouldn't want to make any premature conclusions. However I can imagine a future in which `jsverify` is kept in maintenance mode, focusing on bug fixes / typings / refinements, and users are encouraged to check out `fast-check` for a more actively updated library with a different approach and ongoing feature development.

I am not at all saying that I know enough about either project to declare that that _should_ be how these projects are handled, just floating one option for consideration by those who know this landscape better.

**Comment by phadej** [comment]
> However I can imagine a future in which jsverify is kept in maintenance mode, focusing on bug fixes / typings / refinements

You don’t need to imagine it, that the present state already :) (except, i’m not so good on even merging fixes / typings / etc)

**Comment by phadej** [comment]
> it also has most of its examples written in a different programming language that has little overlap with the target userbase

What you mean by that? Isn’t examples written in JavaScript, or do you mean that most users use TypeScript? Hmm. It’s JS-Verify. If I’d now make TS-Verify, it would look different. JSVerify embraces dynamism of JS.

**Comment by StoneCypher** [comment]
No, what I meant was, you write a lot of your documentations and examples in pseudo-haskell, which most JS people don't actually understand

**Comment by lastmjs** [comment]
Are you talking about the DSL? Yes, personally I've never really used or wanted to use the DSL, seems to just get in the way and make things more confusing

**Comment by glebec** [comment]
@lastmjs I believe that the DSL is part of what @StoneCypher means, but more generally, it's the fact that the documentation is largely based on:

- documentation via types, which many JS-only devs are neither fluent at parsing visually nor familiar with conceptually (e.g. union types, tagged vs. untagged, etc.)
- a custom type syntax that is neither Ramda-like, Flow, TS, etc.
- the use of functional / mathematical concepts like "identity functor" and "monad", functional data structures like `tuple` and `either`, and non-JS language (`void`) for JS values (`undefined`)
- very minimal introduction to arbitraries, generators, and shrinks (familiar to QuickCheck/Haskell programmers, but novel to JS devs)

In short, the documentation is usable for someone with both JS and Haskell experience, but very non-idiomatic / difficult to understand for a JS-only dev. It's not written from the perspective of making the concepts accessible to a non-functional userbase.

The ideal documentation would be both a typed API documentation (which this has, though it can be polished up a bit) but also a fairly thorough guide via examples and use cases, written assuming little to no FP mastery.

**Comment by StoneCypher** [comment]
ya, i mean, i speak js and quickcheck, but i actually do not speak haskell meaningfully (i learned quickcheck in erlang)

the haskell notation has always been my largest barrier

**Comment by StoneCypher** [comment]
Ya, sure, add me as a maintainer, @phadej 

I'll see what I can do

**Comment by dubzzz** [comment]
@phadej @StoneCypher @lastmjs @glebec As I don't exactly know what is the current status for this issue, here is a proposal: I built a tool to ease the migration from jsverify to fast-check - see https://github.com/dubzzz/jsverify-to-fast-check.

This is only a proposal, if the final choice is to keep jsverify maintained there is no need to talk about it anywhere in the documentation. If the final choice is not to take the time to maintain jsverify, maybe we could add a note linking to this project.

**Comment by StoneCypher** [comment]
I'm still up to help maintain, like Oleg originally asked

**Comment by mroderick** [comment]
Reading the documentation of `fast-check`, it is quite clear that this is (has become?) a TypeScript centric library. I deduce this from the fact that most of the documentation is written in TypeScript.

There are still many people who have no interest in using TypeScript, and for whom reading the TypeScript based documentation is awkward at best.

I think there is still a place for a **JavaScript centric** property based testing library, and would love to see JSVerify revitalised with new documentation that uses examples that assumes no prior knowledge of Haskell or TypeScript.

How can we get this going again?
